gmr/gmm.py

```diff
@@ -1,8 +1,89 @@
 import numpy as np
+from scipy.spatial.distance import cdist, pdist
 from .utils import check_random_state
 from .mvn import MVN
 
 
+def kmeansplusplus_initialization(X, n_components, random_state=None):
+    """k-means++ initialization for centers of a GMM.
+
+    Initialization of GMM centers before expectation maximization (EM).
+    The first center is selected uniformly random. Subsequent centers are
+    sampled from the data with probability proportional to the squared
+    distance to the closest center.
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+        Samples from the true distribution.
+
+    n_components : int (> 0)
+        Number of MVNs that compose the GMM.
+
+    random_state : int or RandomState, optional (default: global random state)
+        If an integer is given, it fixes the seed. Defaults to the global numpy
+        random number generator.
+
+    Returns
+    -------
+    initial_means : array, shape (n_components, n_features)
+        Initial means
+    """
+    if n_components <= 0:
+        raise ValueError("Only n_components > 0 allowed.")
+    if n_components > len(X):
+        raise ValueError(
+            "More components (%d) than samples (%d) are not allowed."
+            % (n_components, len(X)))
+
+    random_state = check_random_state(random_state)
+
+    all_indices = np.arange(len(X))
+    selected_centers = [random_state.choice(all_indices, size=1).tolist()[0]]
+    while len(selected_centers) < n_components:
+        centers = np.atleast_2d(X[np.array(selected_centers, dtype=int)])
+        squared_distances = cdist(X, centers, metric="sqeuclidean")
+        selection_probability = squared_distances.max(axis=1)
+        selection_probability[np.array(selected_centers, dtype=int)] = 0.0
```

Comment:
> I don't understand why this is necessary

If you mean this particular line: to avoid sampling the same example more than once.

> In a more general matter, why not using the Kmeans fro sklearn for initialization?

That would add another dependency. I'm thinking about mentioning that gmr is fully compatible to sklearn though. I will probably add a new example script soon and mention it in the readme.

You can even use sklearn's Dirichlet process GMM to "initialize" priors, means, and covariances.

[k-means++ initialization](https://en.wikipedia.org/wiki/K-means%2B%2B#Improved_initialization_algorithm) is a well-known initialization method for k-means and GMMs. It requires less computational power than k-means, which might be an advantage for large datasets. Nevertheless, I don't know whether k-means initialization + less EM iterations or k-means++ initialization + more EM iterations is more expensive... I'll let the user decide.

Reply:
For the first point: My bad, forgot about the loop.
For the second point: From what I remember EM has more parameters (N*(N+1)/2) to optimize than Kmeans thus the complexity will be higher I guess?

Do you think using hard K-means would improve the model?

Reply:
Standard k-means doesn't have a covariance. Using k-means for initialization is certainly a good way to initialize the means of a GMM. sklearn does it [by default](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/mixture/_base.py#L145).

Reply:
Yes my bad, I thought you used Soft K-means for your above picture.